mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,662 @@
+    Metrics,
+    TrainState,
+    Transition,
+)
+from mava.types import Observation
+from mava.utils import make_env as environments
+from mava.utils.checkpointing import Checkpointer
+from mava.utils.jax import unreplicate_batch_dim, unreplicate_n_dims
+from mava.utils.logger import LogEvent, MavaLogger
+from mava.wrappers import episode_metrics
+
+
+def init(
+    cfg: DictConfig,
+) -> Tuple[
+    Tuple[Environment, Environment],
+    RecQNetwork,
+    optax.GradientTransformation,
+    TrajectoryBuffer,
+    LearnerState,
+    MavaLogger,
+    PRNGKey,
+]:
+    """Initialize system by creating the envs, networks etc.
+
+    Args:
+        cfg: System configuration.
+
+    Returns:
+        Tuple containing:
+            Tuple[Environment, Environment]: The environment and evaluation environment.
+            RecQNetwork: Recurrent Q network
+            optax.GradientTransformation: Optimiser for RecQNetwork.
+            TrajectoryBuffer: The replay buffer.
+            LearnerState: The initial learner state.
+            MavaLogger: The logger.
+            PRNGKey: The random key.
+    """
+    logger = MavaLogger(cfg)
+
+    key = jax.random.PRNGKey(cfg.system.seed)
+    devices = jax.devices()
+
+    def replicate(x: Any) -> Any:
+        """First replicate the update batch dim then put on devices."""
+        x = jax.tree_map(lambda y: jnp.broadcast_to(y, (cfg.system.update_batch_size, *y.shape)), x)
+        return jax.device_put_replicated(x, devices)
+
+    env, eval_env = environments.make(cfg)
+
+    action_dim = env.action_dim
+    num_agents = env.num_agents
+
+    key, q_key = jax.random.split(key, 2)
+    # Shape legend:
+    # T: Time
+    # B: Batch
+    # A: Agent
+    # Make dummy inputs to init recurrent Q network -> need shape (T, B, A, ...)
+    init_obs = env.observation_spec().generate_value()  # (A, ...)
+    init_obs_batched = jax.tree_util.tree_map(
+        lambda x: jnp.repeat(x[jnp.newaxis, ...], cfg.arch.num_envs, axis=0),
+        init_obs,
+    )  # (B, A, ...)
+    init_obs_batched = jax.tree_util.tree_map(
+        lambda x: x[jnp.newaxis, ...], init_obs_batched
+    )  # add time dim (1, B, A, ...)
+    init_done = jnp.zeros((1, cfg.arch.num_envs, 1), dtype=bool)  # (1, B, 1)
+    init_x = (init_obs_batched, init_done)  # pack the RNN dummy inputs
+    init_hidden_state = ScannedRNN.initialize_carry(
+        (cfg.arch.num_envs, num_agents), cfg.network.hidden_state_dim
+    )  # (B, A, ...)
+
+    # Making recurrent Q network
+    pre_torso = hydra.utils.instantiate(cfg.network.q_network.pre_torso)
+    post_torso = hydra.utils.instantiate(cfg.network.q_network.post_torso)
+    q_net = RecQNetwork(
+        pre_torso,
+        post_torso,
+        action_dim,
+        cfg.network.hidden_state_dim,
+    )
+    q_params = q_net.init(q_key, init_hidden_state, init_x)  # epsilon defaults to 0
+    q_target_params = q_net.init(q_key, init_hidden_state, init_x)  # ensure parameters are separate
```

Comment:
This will generate the same param values for online and target net, just checking that this is desired?

Reply:
Yes! The other way I'd do this is with a deepcopy; we just need to make doubly sure that we don't just have pointers pointing to the same parameters.